pkg/odo/cli/service/ui/ui.go

```diff
@@ -0,0 +1,307 @@
+func getServiceClassNames(stringMap map[string]scv1beta1.ClusterServiceClass) (keys []string) {
+	keys = make([]string, len(stringMap))
+
+	i := 0
+	for k := range stringMap {
+		keys[i] = k
+		i++
+	}
+
+	sort.Strings(keys)
+
+	return keys
+}
+
+// handleError handles UI-related errors, in particular useful to gracefully handle ctrl-c interrupts gracefully
+func handleError(err error) {
+	if err != nil {
+		if err == terminal.InterruptErr {
+			os.Exit(-1)
+		} else {
+			glog.V(4).Infof("Encountered an error processing prompt: %v", err)
+		}
+	}
+}
+
+// SelectPlanNameInteractively lets the user to select the plan name from possible options, specifying which text should appear
+// in the prompt
+func SelectPlanNameInteractively(plans map[string]scv1beta1.ClusterServicePlan, promptText string) (plan string) {
+	prompt := &survey.Select{
+		Message: promptText,
+		Options: GetServicePlanNames(plans),
+	}
+	err := survey.AskOne(prompt, &plan, nil)
+	handleError(err)
+	return plan
+}
+
+// EnterServiceNameInteractively lets the user enter the name of the service instance to create, defaulting to the provided
+// default value and specifying both the prompt text and validation function for the name
+func EnterServiceNameInteractively(defaultValue, promptText string, validateName Validator) (serviceName string) {
+	// if only one arg is given, ask to Name the service providing the class Name as default
+	instancePrompt := &survey.Input{
+		Message: promptText,
+		Default: defaultValue,
+	}
+	err := survey.AskOne(instancePrompt, &serviceName, survey.Validator(validateName))
+	handleError(err)
+	return serviceName
+}
+
+// SelectClassInteractively lets the user select target service class from possible options, first filtering by categories then
+// by class name
+func SelectClassInteractively(classesByCategory map[string][]scv1beta1.ClusterServiceClass) (class scv1beta1.ClusterServiceClass, serviceType string) {
+	var category string
+	prompt := &survey.Select{
+		Message: "Which kind of service do you wish to create",
+		Options: getServiceClassesCategories(classesByCategory),
+	}
+	err := survey.AskOne(prompt, &category, survey.Required)
+	handleError(err)
+
+	classes := getServiceClassMap(classesByCategory[category])
+
+	// make a new displayClassInfo function available to survey templates to be able to add class information to the display
+	displayClassInfo := "displayClassInfo"
+	core.TemplateFuncs[displayClassInfo] = func(index int, pageEntries []string) string {
+		if index >= 0 && len(pageEntries) > index {
+			selected := pageEntries[index]
+			class := classes[selected]
+			return fmt.Sprintf("Name: %s\nDescription: %s\nLong: %s", class.GetExternalName(), class.GetDescription(), getLongDescription(class))
+		}
+		return "No matching entry"
+	}
+	defer delete(core.TemplateFuncs, displayClassInfo)
+
+	// record original template and defer restoring it once done
```

Comment:
it might be a good idea to add a comment why the following block is commented out